### PR TITLE
Tree Tab 2.0

### LINKED ISSUE #100: Adapt tree view for SNOMED CT codelists
The tree view of a codelist is intended all the codelist's concepts in the context of the coding system's hierarchy.  [Here's an example of a CTV3 codelist](https://codelists.opensafely.org/codelist/opensafely/aplastic-anaemia/2020-04-24/#tree).  However the current approach does not work for SNOMED CT codelists.

Most coding system hierarchies allow a concept to have multiple parents, and so there are often multiple paths to a concept from the root of the hierarchy.  That is, the hierarchies are directed acyclic graphs, not tree.  These can be difficult to display meaningfully.

We cope with this by showing each concept in the codelist multiple times, once per path from the root.  In other words, we convert the DAG to a tree.  In the example above, Adult Fanconi Syndrome appears 3 times.

This is practical for CTV3 codelists, because relatively few CTV3 concepts have multiple parents.  However, the SNOMED CT hierarchy is much richer and most SNOMED CT concepts have multiple parents.  This means that there can be hundred of paths from the root to any given concept.  For instance, there are 507 paths from the root to "Acute severe exacerbation of mild persistent allergic asthma co-occurrent with allergic rhinitis (disorder)".

This makes viewing the tree difficult, as a concept can also appear hundreds of times.  Additionally, it makes it expensive to compute, transfer, and render the tree.  (I don't have numbers, but when I tried this out, I recall it taking tens of seconds to compute the tree structure; tens of seconds to render the template in Django; and minutes to render the HTML in the browser.  All of these could be optimised, but there seemed little point if the end result is unusable.)

So for now, we have disabled the tree view for SNOMED CT codelists, which limits the usefulness of the page.

A proposed solution is that instead of showing a single tree whose root is the root of the hierarchy, we show several trees -- one for each concept in the codelist that has no ancestors in the codelist.  These should be grouped together by top-level concept.  This would mean that "Acute severe exacerbation of mild persistent allergic asthma co-occurrent with allergic rhinitis (disorder)" appears only 15 times in the asthma codelist tree.

This is the approach I've used in the prototype codelist builder, where it seems to work:

image

The structure of the tree, and any HTML, could be precomputed/cached for performance if necessary

(The page could be made a bit nicer visually while we're at it!)

### LINKED ISSUE #165: Add collapse/expand to Tree tab
Should match behaviour of builder interface.  Depends on #100.

### PR DESCRIPTION
**Pull request by ghickman**
This rebuilds the Codelist version page's Tree tab with a modified version of the CodelistBuilder component.  It turned out that only `TermAndCode` was needed from CodelistBuilder so I've left the rest of the sub components where they are.

This also pulls a couple of `tree_tables`-related structures into the coding system specific modules (so we can support CTV3 and SNOMED).  Most notbably, concept codes grouped by type are now generated in the coding system (since CTV3 codes can have more than one group) and passed into `tree_tables`.

These changes brought up the single code Hierarchy edge case (#193), which in turn brought up the three view tests (checking redirects work as expected) that broke due to this work.  I've reduced their scope to only check they are redirecting to the correct location so they're less brittle in the future.

**Aplastic Anaemia (SNOMED)**

**Aplastic Anaemia**

Fixes #100 
Fixes #165

### CODE REVIEW

**Review by inglesp — CHANGES_REQUESTED** [review]
Lots of good stuff here, but the tree view is no longer showing whether a code is in the codelist or not!

File: coding_systems/ctv3/coding_system.py [hunk 1]
```diff
@@ -63,3 +63,32 @@ def descendant_relationships(codes):
...
+    """
+    types = hierarchy.child_map[root]
+    concepts = Concept.objects.filter(read_code__in=types).prefetch_related("terms")
+    terms_by_type = {c.read_code: c.preferred_term() for c in concepts}
+
+    for code in codes:
+        type_nodes = hierarchy.ancestors(code) & types
+
+        # TODO: can we pass in codes bucketed by type to tree_tables?
+        first = list(type_nodes)[0]
```
[the remark this answers is not in the document]
Comment (ghickman) [reply coding_systems/ctv3/coding_system.py]: Quite possibly (sets are unordered IIRC?), but this went away in a future commit 🙂

File: static/src/js/tree/index.jsx [hunk 2]
```diff
@@ -0,0 +1,20 @@
+"use strict";
+
+import ReactDOM from "react-dom";
+// Although React is not used in this module, if this line is removed, there is
+// a runtime error in codelistbuilder.jsx.
```
[the remark this answers is not in the document]
Comment (ghickman) [reply static/src/js/tree/index.jsx:5]: In the `ReactDOM.render()` call we pass a JSX element (`<Tree>` here), that's syntactic sugar for `React.createElement()` so we have to have the `React` import.

**Comment by ghickman** [comment]
Black, blue, and underlined codes reinstated.  I've also fixed a bug with alignment of the descendant toggles and pipes.

I removed the `TermAndCode` component in favour of separate components for each of its constituent parts.  Both the builder and tree want to display the items like this:

```
[ N pipes ] [ +/- /? ] [ Term ] [ Code ]
```

where the `+/-/?` section is Builder specific and the `Term` needs to be optionally linked.  So by breaking out these parts both the `CodelistBuilder` and `Tree` components can compose them as necesary.

### Codelist Builder


### Tree (CTV3)


### Tree (SNOMED)

File: templates/codelists/version.html [hunk 3]
```diff
@@ -344,6 +344,8 @@ <h5 class="modal-title" id="publish-version-label">Publish Version</h5>
 </script>
 
 {{ child_map|json_script:"child-map" }}
+{{ code_to_url|json_script:"codeToURL" }}
+{{ codes_in_definition|json_script:"definitionCodes" }}
```
Comment (inglesp) [inline templates/codelists/version.html:348]: Can this be `codesInDefinition`?

File: static/src/js/tree/Tree.jsx [hunk 4]
```diff
@@ -30,6 +30,28 @@ class Tree extends React.Component {
...
   }
 
+  renderTerm(code, term) {
+    const element = <Term term={term} />;
+
+    const inDefinition = this.props.definitionCodes.includes(code);
+
+    // conditionally wrap Term in an <href> element if there's a URL for it. We
+    // only want to link to codes included in the codelist so some codes will
+    // stay unlinked.
```
Comment (inglesp) [inline static/src/js/tree/Tree.jsx:40]: > We only want to link to codes included in the codelist so some codes will stay unlinked.

Why do we want this?

Looks like this might be a result of building up `code_to_url` in the view.  I don't think we need to do this, since the URL can be determined from the coding system ID (which the JS doesn't yet have) and the concept code (which it does have).  As such, I think it's worth hard coding the URL structure in the JS.

**Review by inglesp — APPROVED** [review]
Looks good.  A minor tweak suggested.

File: templates/codelists/version.html (hunk 3 above)
Comment (ghickman) [reply templates/codelists/version.html:348]: The JS variable?

File: static/src/js/tree/Tree.jsx (hunk 4 above)
Comment (ghickman) [reply static/src/js/tree/Tree.jsx:40]: > Why do we want this?

I might have misunderstood something here, I'd assumed we only wanted the included codes linked.  Should I have linked all the codes?

Comment (inglesp) [reply static/src/js/tree/Tree.jsx:40]: That matches the existing behaviour.

File: templates/codelists/version.html (hunk 3 above)
Comment (inglesp) [reply templates/codelists/version.html:348]: Yeah -- that would make it match the pattern in all the other `json_script`s.

File: static/src/js/tree/Tree.jsx (hunk 4 above)
Comment (ghickman) [reply static/src/js/tree/Tree.jsx]: Fixed in 67abb8a

File: templates/codelists/version.html (hunk 3 above)
Comment (ghickman) [reply templates/codelists/version.html]: Fixed in 59f3365
